File: backend/controllers/GioHangController.js
const { GioHang, BienTheSanPham, SanPham, HinhAnhSanPham } = require("../models");

// 1. Lấy giỏ hàng của người dùng đã đăng nhập
exports.getCart = async (req, res) => {
  try {
    const tai_khoan_id = req.user.id;
    const items = await GioHang.findAll({
      where: { tai_khoan_id },
      include: [
        {
          model: BienTheSanPham,
          as: "bien_the",
          include: [
            {
              model: SanPham,
              as: "san_pham",
              include: [
                {
                  model: HinhAnhSanPham,
                  as: "hinh_anh",
                },
              ],
            },
          ],
        },
      ],
    });

    const formattedCart = items
      .map((item) => {
        const bt = item.bien_the;
        if (!bt) return null;
        const sp = bt.san_pham;
        if (!sp) return null;

        // Xác định ảnh phù hợp: ảnh cụ thể của biến thể hoặc ảnh chính của sản phẩm
        let hinhAnh = null;
        if (sp.hinh_anh && sp.hinh_anh.length > 0) {
          const variantImg = sp.hinh_anh.find((img) => img.bien_the_id === bt.id);
          if (variantImg) {
            hinhAnh = variantImg.url_anh;
          } else {
            const mainImg =
              sp.hinh_anh.find((img) => img.la_anh_chinh) || sp.hinh_anh[0];
            hinhAnh = mainImg?.url_anh;
          }
        }

        return {
          id: sp.id,
          variantId: bt.id,
          ten_san_pham: sp.ten_san_pham,
          hinh_anh: hinhAnh || "",
          gia_ban: Number(bt.gia_ban || bt.gia_goc || 0),
          dung_luong: bt.dung_luong || "",
          mau_sac: bt.mau_sac || "",
          ram: bt.ram || "",
          sku: bt.sku || "",
          so_luong: item.so_luong,
          selected: true, // mặc định được chọn trong trang giỏ hàng
        };
      })
      .filter(Boolean);

    res.json(formattedCart);
  } catch (error) {
    console.error("Lỗi lấy giỏ hàng:", error);
    res.status(500).json({ message: "Lỗi server khi lấy giỏ hàng" });
  }
};

// 2. Thêm mặt hàng vào giỏ hàng DB
exports.addToCart = async (req, res) => {
  try {
    const tai_khoan_id = req.user.id;
    const { bien_the_id, so_luong } = req.body;

    if (!bien_the_id || !so_luong) {
      return res
        .status(400)
        .json({ message: "Thiếu thông tin biến thể hoặc số lượng!" });
    }

    const bt = await BienTheSanPham.findByPk(bien_the_id);
    if (!bt) {
      return res.status(404).json({ message: "Biến thể sản phẩm không tồn tại!" });
    }

    // Kiểm tra đã có trong giỏ chưa
    const existing = await GioHang.findOne({
      where: { tai_khoan_id, bien_the_id },
    });

    const newQty = existing
      ? existing.so_luong + Number(so_luong)
      : Number(so_luong);

    if (newQty > bt.ton_kho) {
      return res.status(400).json({
        message: `Không thể thêm! Bạn đã có ${
          existing ? existing.so_luong : 0
        } sản phẩm trong giỏ hàng. Tồn kho khả dụng là ${bt.ton_kho}.`,
      });
    }

    if (existing) {
      existing.so_luong = newQty;
      await existing.save();
    } else {
      await GioHang.create({
        tai_khoan_id,
        bien_the_id,
        so_luong: newQty,
      });
    }

    res.json({ success: true, message: "Đã thêm vào giỏ hàng thành công!" });
  } catch (error) {
    console.error("Lỗi thêm vào giỏ hàng:", error);
    res.status(500).json({ message: "Lỗi server khi thêm vào giỏ hàng" });
  }
};

// 3. Cập nhật số lượng của một sản phẩm trong giỏ hàng DB
exports.updateQuantity = async (req, res) => {
  try {
    const tai_khoan_id = req.user.id;
    const { bien_the_id, so_luong } = req.body;

    if (!bien_the_id || so_luong === undefined) {
      return res.status(400).json({ message: "Thiếu thông tin!" });
    }

    const targetQty = Number(so_luong);
    if (targetQty <= 0) {
      await GioHang.destroy({ where: { tai_khoan_id, bien_the_id } });
      return res.json({ success: true, message: "Đã xóa sản phẩm khỏi giỏ hàng!" });
    }

    const bt = await BienTheSanPham.findByPk(bien_the_id);
    if (!bt) {
      return res.status(404).json({ message: "Biến thể không tồn tại!" });
    }

    if (targetQty > bt.ton_kho) {
      return res
        .status(400)
        .json({ message: `Chỉ còn ${bt.ton_kho} sản phẩm khả dụng!` });
    }

    const [item, created] = await GioHang.findOrCreate({
      where: { tai_khoan_id, bien_the_id },
      defaults: { so_luong: targetQty },
    });

    if (!created) {
      item.so_luong = targetQty;
      await item.save();
    }

    res.json({ success: true, message: "Đã cập nhật số lượng!" });
  } catch (error) {
    console.error("Lỗi cập nhật số lượng:", error);
    res.status(500).json({ message: "Lỗi server khi cập nhật số lượng" });
  }
};

// 4. Xóa sản phẩm khỏi giỏ hàng DB
exports.removeFromCart = async (req, res) => {
  try {
    const tai_khoan_id = req.user.id;
    const { bien_the_id } = req.params;

    await GioHang.destroy({ where: { tai_khoan_id, bien_the_id } });
    res.json({ success: true, message: "Đã xóa sản phẩm khỏi giỏ hàng!" });
  } catch (error) {
    console.error("Lỗi xóa giỏ hàng:", error);
    res.status(500).json({ message: "Lỗi server khi xóa sản phẩm" });
  }
};

// 5. Đồng bộ gộp giỏ hàng từ Local vào Database khi đăng nhập
exports.syncCart = async (req, res) => {
  try {
    const tai_khoan_id = req.user.id;
    const { items } = req.body;

    if (Array.isArray(items) && items.length > 0) {
      for (const item of items) {
        const { variantId, so_luong } = item;
        if (!variantId || !so_luong) continue;

        const bt = await BienTheSanPham.findByPk(variantId);
        if (!bt) continue;

        const existing = await GioHang.findOne({
          where: { tai_khoan_id, bien_the_id: variantId },
        });

        // Gộp số lượng
        let mergedQty = existing
          ? existing.so_luong + Number(so_luong)
          : Number(so_luong);

        // Giới hạn trong số lượng tồn kho khả dụng
        if (mergedQty > bt.ton_kho) {
          mergedQty = bt.ton_kho;
        }

        if (mergedQty > 0) {
          if (existing) {
            existing.so_luong = mergedQty;
            await existing.save();
          } else {
            await GioHang.create({
              tai_khoan_id,
              bien_the_id: variantId,
              so_luong: mergedQty,
            });
          }
        }
      }
    }

    // Trả về giỏ hàng sau khi gộp đồng bộ xong
    return exports.getCart(req, res);
  } catch (error) {
    console.error("Lỗi đồng bộ giỏ hàng:", error);
    res.status(500).json({ message: "Lỗi server khi đồng bộ giỏ hàng" });
  }
};

// 6. Xóa các sản phẩm được chọn khỏi giỏ hàng DB sau khi thanh toán thành công
exports.clearSelected = async (req, res) => {
  try {
    const tai_khoan_id = req.user.id;
    const { variantIds } = req.body;

    if (!Array.isArray(variantIds) || variantIds.length === 0) {
      return res
        .status(400)
        .json({ message: "Thiếu danh sách mặt hàng cần xóa!" });
    }

    await GioHang.destroy({
      where: {
        tai_khoan_id,
        bien_the_id: variantIds,
      },
    });

    res.json({ success: true, message: "Đã dọn dẹp các sản phẩm đã thanh toán!" });
  } catch (error) {
    console.error("Lỗi dọn dẹp giỏ hàng sau thanh toán:", error);
    res.status(500).json({ message: "Lỗi server khi dọn dẹp giỏ hàng" });
  }
};

File: backend-csharp/Controllers/GioHangController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using WebFashion.Api.Models;

namespace WebFashion.Api.Controllers
{
    [Route("api/cart")]
    [ApiController]
    [Authorize]
    public class GioHangController : ControllerBase
    {
        private readonly AppDbContext _context;

        public GioHangController(AppDbContext context)
        {
            _context = context;
        }

        private long GetUserId()
        {
            var idClaim = User.FindFirst("id")?.Value;
            if (string.IsNullOrEmpty(idClaim) || !long.TryParse(idClaim, out var userId))
            {
                throw new UnauthorizedAccessException("Chưa xác thực người dùng!");
            }
            return userId;
        }

        // 1. LẤY GIỎ HÀNG CỦA NGƯỜI DÙNG ĐÃ ĐĂNG NHẬP
        // GET: api/cart
        [HttpGet]
        public async Task<IActionResult> GetCart()
        {
            try
            {
                var userId = GetUserId();
                var items = await _context.GioHangs
                    .Where(g => g.TaiKhoanId == userId)
                    .Include(g => g.BienThe)
                        .ThenInclude(bt => bt.SanPham)
                            .ThenInclude(sp => sp.HinhAnhSanPhams)
                    .ToListAsync();

                var formattedCart = items.Select(item =>
                {
                    var bt = item.BienThe;
                    if (bt == null) return null;
                    var sp = bt.SanPham;
                    if (sp == null) return null;

                    // Match image: specific variant image or main product image
                    string hinhAnh = "";
                    if (sp.HinhAnhSanPhams != null && sp.HinhAnhSanPhams.Count > 0)
                    {
                        var variantImg = sp.HinhAnhSanPhams.FirstOrDefault(img => img.BienTheId == bt.Id);
                        if (variantImg != null)
                        {
                            hinhAnh = variantImg.UrlAnh;
                        }
                        else
                        {
                            var mainImg = sp.HinhAnhSanPhams.FirstOrDefault(img => img.LaAnhChinh) ?? sp.HinhAnhSanPhams.First();
                            hinhAnh = mainImg.UrlAnh;
                        }
                    }

                    return new
                    {
                        id = sp.Id,
                        variantId = bt.Id,
                        ten_san_pham = sp.TenSanPham,
                        hinh_anh = hinhAnh,
                        gia_ban = (double)bt.GiaBan,
                        dung_luong = bt.DungLuong ?? "",
                        mau_sac = bt.MauSac ?? "",
                        ram = bt.Ram ?? "",
                        sku = bt.Sku,
                        so_luong = item.SoLuong,
                        selected = true // Default true for client checkout selectors
                    };
                })
                .Where(x => x != null)
                .ToList();

                return Ok(formattedCart);
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi lấy giỏ hàng: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server khi lấy giỏ hàng" });
            }
        }

        // 2. THÊM MẶT HÀNG VÀO GIỎ HÀNG DB
        // POST: api/cart
        [HttpPost]
        public async Task<IActionResult> AddToCart([FromBody] JsonElement body)
        {
            try
            {
                var userId = GetUserId();
                
                if (!body.TryGetProperty("bien_the_id", out var vtProp) || 
                    !body.TryGetProperty("so_luong", out var qtyProp))
                {
                    return BadRequest(new { message = "Thiếu thông tin biến thể hoặc số lượng!" });
                }

                long bienTheId = vtProp.GetInt64();
                int soLuong = qtyProp.GetInt32();

                if (bienTheId <= 0 || soLuong <= 0)
                {
                    return BadRequest(new { message = "Thông tin không hợp lệ!" });
                }

                var bt = await _context.BienTheSanPhams.FindAsync(bienTheId);
                if (bt == null)
                {
                    return NotFound(new { message = "Biến thể sản phẩm không tồn tại!" });
                }

                // Check if already in cart
                var existing = await _context.GioHangs
                    .FirstOrDefaultAsync(g => g.TaiKhoanId == userId && g.BienTheId == bienTheId);

                int newQty = existing != null ? existing.SoLuong + soLuong : soLuong;

                if (newQty > bt.TonKho)
                {
                    return BadRequest(new
                    {
                        message = $"Không thể thêm! Bạn đã có {(existing != null ? existing.SoLuong : 0)} sản phẩm trong giỏ hàng. Tồn kho khả dụng là {bt.TonKho}."
                    });
                }

                if (existing != null)
                {
                    existing.SoLuong = newQty;
                    existing.UpdatedAt = DateTime.UtcNow;
                }
                else
                {
                    var newCartItem = new GioHang
                    {
                        TaiKhoanId = userId,
                        BienTheId = bienTheId,
                        SoLuong = newQty,
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow
                    };
                    _context.GioHangs.Add(newCartItem);
                }

                await _context.SaveChangesAsync();
                return Ok(new { success = true, message = "Đã thêm vào giỏ hàng thành công!" });
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi thêm vào giỏ hàng: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server khi thêm vào giỏ hàng" });
            }
        }

        // 3. CẬP NHẬT SỐ LƯỢNG MỘT MẶT HÀNG
        // PUT: api/cart
        [HttpPut]
        public async Task<IActionResult> UpdateQuantity([FromBody] JsonElement body)
        {
            try
            {
                var userId = GetUserId();

                if (!body.TryGetProperty("bien_the_id", out var vtProp) || 
                    !body.TryGetProperty("so_luong", out var qtyProp))
                {
                    return BadRequest(new { message = "Thiếu thông tin!" });
                }

                long bienTheId = vtProp.GetInt64();
                int targetQty = qtyProp.GetInt32();

                if (targetQty <= 0)
                {
                    var itemToDelete = await _context.GioHangs
                        .FirstOrDefaultAsync(g => g.TaiKhoanId == userId && g.BienTheId == bienTheId);
                    
                    if (itemToDelete != null)
                    {
                        _context.GioHangs.Remove(itemToDelete);
                        await _context.SaveChangesAsync();
                    }
                    return Ok(new { success = true, message = "Đã xóa sản phẩm khỏi giỏ hàng!" });
                }

                var bt = await _context.BienTheSanPhams.FindAsync(bienTheId);
                if (bt == null)
                {
                    return NotFound(new { message = "Biến thể không tồn tại!" });
                }

                if (targetQty > bt.TonKho)
                {
                    return BadRequest(new { message = $"Chỉ còn {bt.TonKho} sản phẩm khả dụng!" });
                }

                var item = await _context.GioHangs
                    .FirstOrDefaultAsync(g => g.TaiKhoanId == userId && g.BienTheId == bienTheId);

                if (item == null)
                {
                    item = new GioHang
                    {
                        TaiKhoanId = userId,
                        BienTheId = bienTheId,
                        SoLuong = targetQty,
                        CreatedAt = DateTime.UtcNow,
                        UpdatedAt = DateTime.UtcNow
                    };
                    _context.GioHangs.Add(item);
                }
                else
                {
                    item.SoLuong = targetQty;
                    item.UpdatedAt = DateTime.UtcNow;
                }

                await _context.SaveChangesAsync();
                return Ok(new { success = true, message = "Đã cập nhật số lượng!" });
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi cập nhật số lượng: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server khi cập nhật số lượng" });
            }
        }

        // 4. XÓA MỘT SẢN PHẨM KHỎI GIỎ HÀNG
        // DELETE: api/cart/{bien_the_id}
        [HttpDelete("{bien_the_id}")]
        public async Task<IActionResult> RemoveFromCart(long bien_the_id)
        {
            try
            {
                var userId = GetUserId();
                var item = await _context.GioHangs
                    .FirstOrDefaultAsync(g => g.TaiKhoanId == userId && g.BienTheId == bien_the_id);

                if (item != null)
                {
                    _context.GioHangs.Remove(item);
                    await _context.SaveChangesAsync();
                }

                return Ok(new { success = true, message = "Đã xóa sản phẩm khỏi giỏ hàng!" });
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi xóa giỏ hàng: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server khi xóa sản phẩm" });
            }
        }

        // 5. ĐỒNG BỘ GỘP GIỎ HÀNG (Local storage to DB on login)
        // POST: api/cart/sync
        [HttpPost("sync")]
        public async Task<IActionResult> SyncCart([FromBody] JsonElement body)
        {
            try
            {
                var userId = GetUserId();

                if (body.TryGetProperty("items", out var itemsProp) && itemsProp.ValueKind == JsonValueKind.Array)
                {
                    foreach (var item in itemsProp.EnumerateArray())
                    {
                        if (!item.TryGetProperty("variantId", out var vIdProp) || 
                            !item.TryGetProperty("so_luong", out var qtyProp))
                        {
                            continue;
                        }

                        long variantId = vIdProp.GetInt64();
                        int qty = qtyProp.GetInt32();

                        if (variantId <= 0 || qty <= 0) continue;

                        var bt = await _context.BienTheSanPhams.FindAsync(variantId);
                        if (bt == null) continue;

                        var existing = await _context.GioHangs
                            .FirstOrDefaultAsync(g => g.TaiKhoanId == userId && g.BienTheId == variantId);

                        int mergedQty = existing != null ? existing.SoLuong + qty : qty;

                        // Bound by stock
                        if (mergedQty > bt.TonKho)
                        {
                            mergedQty = bt.TonKho;
                        }

                        if (mergedQty > 0)
                        {
                            if (existing != null)
                            {
                                existing.SoLuong = mergedQty;
                                existing.UpdatedAt = DateTime.UtcNow;
                            }
                            else
                            {
                                var newCartItem = new GioHang
                                {
                                    TaiKhoanId = userId,
                                    BienTheId = variantId,
                                    SoLuong = mergedQty,
                                    CreatedAt = DateTime.UtcNow,
                                    UpdatedAt = DateTime.UtcNow
                                };
                                _context.GioHangs.Add(newCartItem);
                            }
                        }
                    }

                    await _context.SaveChangesAsync();
                }

                // Return active mapped cart items directly after sync completes
                return await GetCart();
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi đồng bộ giỏ hàng: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server khi đồng bộ giỏ hàng" });
            }
        }

        // 6. XÓA CÁC SẢN PHẨM ĐÃ THANH TOÁN KHỎI GIỎ HÀNG DB
        // POST: api/cart/clear-selected
        [HttpPost("clear-selected")]
        public async Task<IActionResult> ClearSelected([FromBody] JsonElement body)
        {
            try
            {
                var userId = GetUserId();

                if (!body.TryGetProperty("variantIds", out var idsProp) || idsProp.ValueKind != JsonValueKind.Array)
                {
                    return BadRequest(new { message = "Thiếu danh sách mặt hàng cần xóa!" });
                }

                var variantIds = new List<long>();
                foreach (var idEl in idsProp.EnumerateArray())
                {
                    variantIds.Add(idEl.GetInt64());
                }

                if (variantIds.Count == 0)
                {
                    return BadRequest(new { message = "Danh sách mặt hàng trống!" });
                }

                var itemsToDelete = await _context.GioHangs
                    .Where(g => g.TaiKhoanId == userId && variantIds.Contains(g.BienTheId))
                    .ToListAsync();

                if (itemsToDelete.Count > 0)
                {
                    _context.GioHangs.RemoveRange(itemsToDelete);
                    await _context.SaveChangesAsync();
                }

                return Ok(new { success = true, message = "Đã dọn dẹp các sản phẩm đã thanh toán!" });
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi dọn dẹp giỏ hàng: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server khi dọn dẹp giỏ hàng" });
            }
        }
    }
}
